Add AEAD interface

Ryanfsdf opened the pull request:
Added the AEAD interface. This largely resembles [ALTSRecordCrypto](https://github.com/grpc/grpc-go/blob/695da821e98cb7a10c41e3f5008aa994560e551a/credentials/alts/internal/conn/record.go#L33).

The differences between this and `ALTSRecordCrypto` include:
- the name of the interface
- `EncryptionOverhead` -> `TagSize`
- addition of `UpdateKey API

matthewstevenson88 commented on security/s2a/crypter/record.go:
```diff
@@ -0,0 +1,16 @@
+package crypter
+
+// S2AAeadCrypter is the interface for gRPC S2A record protocol.
```
Could we change this comment to `is the interface for an AEAD cipher used by the S2A record protocol.`?

matthewstevenson88 commented on security/s2a/crypter/record.go:
```diff
@@ -0,0 +1,16 @@
+	Encrypt(dst, plaintext []byte) ([]byte, error)
+	// Decrypt decrypts ciphertext and verifies the tag (if any). dst and
+	// ciphertext may alias exactly or not at all. To reuse ciphertext's
+	// storage for the decrypted output, use ciphertext[:0] as dst.
```
Can we remove the `To reuse ciphertext's...` sentence for now? We'll have to check whether all implementations of this interface actually satisfy that and if they do we can add  it back in.

matthewstevenson88 commented on security/s2a/crypter/record.go:
```diff
@@ -0,0 +1,16 @@
+
+// S2AAeadCrypter is the interface for gRPC S2A record protocol.
+type S2AAeadCrypter interface {
+	// Encrypt encrypts the plaintext and computes the tag (if any) of dst
```
Please remove `(if any)` and similarly in the comments of `Decrypt`. For our use cases, we will always have a tag.

matthewstevenson88 commented:
Thanks for the PR Ryan! It looks pretty good. Two additional thoughts:

1. On second thought, can we change the directory structure slightly? Can we put this in `security/s2a/internal/crypter`?
2. Can this be in a file called `common.go` instead of `record.go`?

Ryanfsdf commented:
Thanks for the review! I addressed all the review comments, PTAL.

> On second thought, can we change the directory structure slightly? Can we put this in security/s2a/internal/crypter?

Could I know the reason why we want the additional `internal` folder?

matthewstevenson88 commented:
Thanks for the PR Ryan! I approved and added @cesarghali for Go readability.

I think the `internal` folder is a good idea because it allows us to hide all of the APIs that we do not want users to use (we will also use visibility restrictions in our BUILD files to accomplish this). The public APIs that we want users to call will be exposed in `security/s2a`.

cesarghali commented on security/s2a/internal/crypter/common.go:
```diff
@@ -0,0 +1,15 @@
+package crypter
+
+// S2AAeadCrypter is the interface for an AEAD cipher used by the S2A record protocol.
```
Even though there's no line limit in Go, can we try to keep the comments to 80 characters per line? This makes them easy to read :)

Ryanfsdf replied:
Done. Do you suggest the same for code as well?

cesarghali replied:
No. We don't need to impose any line length for the code, just for the comment since they're texts. The idea is if there's a paragraph of comment, and if the lines are long, they would break in a way that would make the comment hard to read.

cesarghali commented on security/s2a/internal/crypter/common.go:
```diff
@@ -0,0 +1,15 @@
+
+// S2AAeadCrypter is the interface for an AEAD cipher used by the S2A record protocol.
+type S2AAeadCrypter interface {
+	// Encrypt encrypts the plaintext and computes the tag of dst
```
Please try to fit as much as possible in 80 chars. For example, some of the comment text on the next line can be moved to the end of this line.

Same below.

cesarghali commented on security/s2a/internal/crypter/common.go:
```diff
@@ -0,0 +1,15 @@
+// S2AAeadCrypter is the interface for an AEAD cipher used by the S2A record protocol.
+type S2AAeadCrypter interface {
+	// Encrypt encrypts the plaintext and computes the tag of dst
+	// and plaintext. dst and plaintext do not overlap.
```
It might be worth clarifying the comment more. I'm assuming you will be using the AEAD object [1] to implement this interface. If that's the case, then the tag is only computed for the plaintext and the result (ciphertext+tag) is appended to the dst. Also, I think dst and plaintext can either fully overlap, or not overlap at all.


[1] https://golang.org/pkg/crypto/cipher/#AEAD

cesarghali replied:
I just saw Matt's comment about the comment. We don't have to go in details of reusing the plaintext buffer, but we should say that dst and plaintext can either fully overlap or not at all.

Ryanfsdf replied:
Do you think this is worth changing in [ALTSRecordCrypto](https://github.com/grpc/grpc-go/blob/695da821e98cb7a10c41e3f5008aa994560e551a/credentials/alts/internal/conn/record.go#L33) as well? Much of the documentation for the interface was taken from the ALTS interface.

cesarghali replied:
That's a good idea. If you would like to send a PR for gRPC-Go to fix the documentation, please ping me directly and I can review it.

cesarghali commented on security/s2a/internal/crypter/common.go:
```diff
@@ -0,0 +1,15 @@
+	// Encrypt encrypts the plaintext and computes the tag of dst
+	// and plaintext. dst and plaintext do not overlap.
+	Encrypt(dst, plaintext []byte) ([]byte, error)
+	// Decrypt decrypts ciphertext and verifies the tag. dst and
```
Please clarify the comment based on
https://golang.org/pkg/crypto/cipher/#AEAD

cesarghali commented on security/s2a/internal/crypter/common.go:
```diff
@@ -0,0 +1,15 @@
+package crypter
```
I'm curious why the file name is `common.go`? Are we planning to use this interface somewhere else?

matthewstevenson88 replied:
This was my suggestion, though I'm open to changing the name. Assuming the record protocol is in a file named `record.go` (like for ALTS), then I don't think this interface should go in `record.go` because the AEAD crypter will be owned by the half connection, not the record protocol. WDYT?

cesarghali replied:
We can call this file crypter. WDYT? I'm assuming we will have several implementations of this interface in separate files in this directlry.

matthewstevenson88 replied:
Would you be opposed to `aeadcrypter.go`, to distinguish it from the package name?

(I fully admit I don't know what the usual naming conventions are in Go, so I defer to you on this one.)

cesarghali replied:
`aeadcrypter.go` works fine with me. Thanks!

Ryanfsdf commented:
Thanks for the reviews! Addressed all the review comments, PTAL.

cesarghali commented on security/s2a/internal/crypter/aeadcrypter.go:
```diff
@@ -0,0 +1,18 @@
+// protocol.
+type S2AAeadCrypter interface {
+	// Encrypt encrypts plaintext, computes the tag of plaintext, and appends
+	// the resulting ciphertext to dst, returning the updated slice. dst and
```
Sorry for the back and forth on this. Let's keep the documentation the same as ALTS, after your latest fix. Basically, we might not want to commit that the result is append to `dst` since we might do something different in some implementations of the interface.
